data_structures/avltree_map.cpp

```diff
@@ -0,0 +1,272 @@
+/**
+ * @file
+ * @brief Unit testings for `AvlTreeMap<K, V>`
+ * @details This file contains the unit testings for `AvlTreeMap<K, V>`
+ *  including insertions, deletions and queries. These tests cover all
+ *  the code that contains the main algorithmic logic.
+ *  If a unit test succeeds, it will print "testName: success",
+ *  otherwise, it will print "testName: failed `[reason]`".
+ * @author [r.ivance](https://github.com/RIvance)
+ * @see avltree_map.hpp
+ */
+
+#include "avltree_map.hpp"
+
+#include <iostream>     /// for IO operations
+#include <set>          /// for `std::set`
+
+struct AssertionFailed : public std::exception
+{
+    const char* reason;
+    explicit AssertionFailed(const char* reason) : reason(reason) {}
+    const char* what() const noexcept override { return reason; }
+};
+
+#define ASSERT_EQUAL(actual, ...) do { \
+    if (!((actual) == (typeof(actual))(__VA_ARGS__))) throw AssertionFailed(#actual" != "#__VA_ARGS__); \
+} while (false)
+
+#define ASSERT_THROW(code, Ex) do { \
+    try { code; } \
+    catch (Ex & e) { break; } \
+    throw AssertionFailed(#Ex" is not thrown"); \
+} while (false)
+
+#define RUN_TEST(test) do { \
+    try { test(); std::cout << #test":\tsucceeded" << std::endl; } \
+    catch (std::exception & ex) { std::cout << #test":\tfailed `" << ex.what() << "`" << std::endl; } \
+} while (false)
+
+template <typename T>
+bool operator==(const std::vector<T> & actual, const std::vector<T> & expected)
+{
+    if (actual.size() != expected.size()) return false;
+    for (int i = 0; i < actual.size(); i++) {
+        if (actual[i] != expected[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+using TestMap = AvlTreeMap<int, int>;
+
+TestMap initWithKeys(const std::vector<int> & keys)
+{
+    TestMap map;
+    for (const auto & key : keys) {
+        map.insert(key, key);
+    }
+    return map;
+}
+
+/**
+ * @brief Test AvlTreeMap<K, V>::insert
```

Comment:
`@returns void` must be added after the brief description of functions without parameters, otherwise, Doxygen won't register them properly.
```suggestion
 * @brief Test AvlTreeMap<K, V>::insert
 * @returns void
```